Expanded documentation about observability in learning queries (#46)
UPPAALModelChecker/docs.uppaal.org: content/language-reference/requirements-specification/_index.md

### Examples
`minE(cost) [<=10] : <> goal`
: learns a strategy that minimizes the expected `cost` expression within `10` time units or when `goal` predicate becomes true given that the **entire** system state is observable.

`maxE(gain) [<=10] : <> goal`
: learns a strategy that maximizes the expected `gain` expression withing `10` time units or when `goal` predicate becomes true given that the **entire** system state is observable.

The `goal` predicate is deprecated, for best results use a predicate which stops together with the simulation bound, like `t>=10`, where `t` is a clock that is never reset.

`minE(cost) [<=10] { i, j } -> { d, f } : <> goal`

> : learns a strategy that minimizes the expected `cost` expression within `10` time units or when `goal` predicate becomes true. Where only the expressions `i`, `j`, `d` and `f` are observable.  The `{..} -> {..}` syntax controls what is observable. By only observing a subset of the system learning times can be significantly reduced. There are two types of observable state expressions: discrete and continuous. Discrete are specified in the first bracket and the continuous in the second: `{discrete expressions} -> {continouos expressions}`. By default the entire state is considered during learning.

Reviewer: ```suggestion
: learns a strategy that minimizes the expected `cost` expression within `10` time units or when `goal` predicate becomes true. Where only the expressions `i`, `j`, `d` and `f` are observable.  The `{..} -> {..}` syntax controls what is observable. 
On one hand, by observing only a partial state learning times can be significantly reduced and the strategy structure simplified. On the other hand, the expected value can be far from optimum if outcomes do not depend on the observed state.
There are two types of observable state expressions: *discrete* and *continuous*. 
The *discrete* are specified in the first bracket and the *continuous* in the second: `{discrete expressions} -> {continuous expressions}`. 
By default the entire state is considered during learning.
```

> **Discrete** expressions will be observable as is i.e the query `minE(cost) [<=10] { i, j } -> { } : <> goal` the will create a strategy by only observing `i` and `j` values.

Reviewer: ```suggestion
**Discrete** expressions are observed as they are, i.e the query `minE(cost) [<=10] { i, j } -> { } : <> goal` creates a strategy by only observing the values of `i` and `j`.
```

> **Continuous** expressions will be discretized during learning using online partition refinement (see [Teaching Stratego to Play Ball](https://vbn.aau.dk/ws/files/378436068/main.pdf)). The query `minE(cost) [<=10] { } -> { d, f } : <> goal` learns a strategy based on the discretized  expressions `d` and `f`.

Reviewer: I am not confident with is meant here
```suggestion
**Continuous** expression values are grouped using online partition refinement (see [Teaching Stratego to Play Ball](https://vbn.aau.dk/ws/files/378436068/main.pdf)). 
The query `minE(cost) [<=10] { } -> { d, f } : <> goal` learns a strategy based on the partitioned values of expressions `d` and `f`.
```

Author: I have a preference towards the discretization wording, but i've improved the sentence according to other comments

Integers, clocks, floating points or even arbitrary expressions can be used in either type of observabilty. However we suggest caution when using floating point numbers or clocks in discrete observability.  

> Process locations will be ignored when specifying observability unless explicitly specified using the expression `Cat.location` and `Mouse.location` where `Cat` and `Mouse` are processes

Reviewer: ```suggestion
Process locations are ignored when specifying observability unless explicitly specified using `location` keyword.
For example `Cat.location` and `Mouse.location` refer to the locations of `Cat` and `Mouse` processes.
```

### Examples
`minE(cost) [<=10] : <> goal`
: learns a strategy that minimizes the expected `cost` value within `10` time units or when `goal` predicate becomes true given that the **entire** system state is observable.

`maxE(gain) [<=10] : <> goal`
: learns a strategy that maximizes the expected `gain` value withing `10` time units or when `goal` predicate becomes true given that the **entire** system state is observable.

The `goal` predicate is deprecated, for best results use a predicate which stops together with the simulation bound, like `t>=10`, where `t` is a clock that is never reset.

`minE(cost) [<=10] { i, j } -> { d, f } : <> goal`

> : learns a strategy that minimizes the expected `cost` value within `10` time units or when `goal` predicate becomes true. Where only the expressions `i`, `j`, `d` and `f` are observable by the learning agent.  The `{..} -> {..}` syntax controls what is observable to the learning agent. Which can drastically reduce learning times by reducing the state space a learning agent has to consider. By default the learning agent will consider the entire state space.  

Reviewer: `learning agent` is undefined (also it does not add any new information, just confuses).
`entire state space` is wrong, should be `entire state`.

> <br>There are two types of observability, Discrete- and Continouos- observability. These are distinquished by what brackets expressions are put into i.e. `{discrete expressions} -> {continouos expressions}`.

Reviewer: ```suggestion
<br>There are two types of observable state expressions: discrete and continuous. Discrete are specified in the first bracket and the continuous in the second: `{discrete expressions} -> {continouos expressions}`.
```

- Dicrete Observability  

> Any expressions here will be observable by the learning agent i.e given the learning query `minE(cost) [<=10] { i, j } -> { } : <> goal` the learning agent will only be able to create a strategy for all combinations of `i` and `j` values.

Reviewer: This does not add any information. Also `all combinations` is misleading: it will construct a mapping which contains only the observed values.

Author: Removed `all combinations` and cut down on bloat.

- Continuous Observability  
Continuous expressions must be discretized to create a strategy, Uppaal will discretize these expressions using online partial refinement, see [Teaching Stratego to Play Ball](https://vbn.aau.dk/ws/files/378436068/main.pdf) for more details. For the query `minE(cost) [<=10] { } -> { d, f } : <> goal` a learning agent will learn a strategy and a discretization of `d` and `f`.

> 

Reviewer: `online partial refinement` is undefined. Also it is very specific (relates to specific learning methods), whereas UPPAAL supports many methods.

Author: I meant `online partition refinement`, but I don't think you can change how continuous expressions are discretized in Uppaal.

I thought citing a paper would suffice as a definition is out of scope for this documentation.
Although the other option is to remove the mention of how we discretize as mentioning it here may only be distracting
